simdev/module/git/clone_stage.py

```diff
@@ -0,0 +1,140 @@
+import logging
+import os
+import textwrap
+
+from git import GitCommandError
+from pydriller import Repository
+from tqdm import tqdm
+
+from simdev.util.pipeline import Pipeline, Stage, PipelineException
+
+
+class AuthorCompound:
+    """
+    A simple structure that embodies both author's email and name
+    """
+
+    def __init__(self, name: str, email: str):
+        self.name = name
+        self.email = email
+
+    def __repr__(self) -> str:
+        """
+        :return: info about a contributor
+        """
+        return "%s <%s>" % (self.name, self.email)
+
+    def __hash__(self) -> int:
+        return hash((self.name, self.email))
+
+    def __eq__(self, o: "AuthorCompound") -> bool:
+        return (self.name, self.email) == (o.name, o.email)
+
+
+class FileContext:
+    """
+    Context that embodies the change inside a file: added and deleted lines
+    """
+
+    def __init__(self, added_lines: int = 0, deleted_lines: int = 0):
+        self.added_lines = added_lines
+        self.deleted_lines = deleted_lines
+
+    @property
+    def changed_lines(self) -> int:
+        return self.added_lines + self.deleted_lines
+
+
+class ContributorContext:
+    """
+    Context that embodies the author and the set of files that are being changed by them per specific repository
+    """
+
+    def __init__(self, author: AuthorCompound):
+        self.author = author
+        self.files: dict[str, FileContext] = {}
+
+    def __repr__(self) -> str:
+        """
+        :return: info about changed files
+        """
+        return "%s: %s" % (repr(self.author), repr(self.files))
+
+
+class RepositoryContext:
+    """
+    Context that embodies a repository alongside with their contributors' contexts and whether this repository is
+    excluded from further consideration or not due to any reasons
+    """
+
+    def __init__(self, url: str):
+        self.url = url
+        self.repository: RepositoryContext | None = None
+        self.contributors: dict[AuthorCompound] = {}
```

Comment:
make it defaultdict - `defaultdict(AuthorCompound)` instead of `{}`
the same applies to files and other `dict` where you call `setdefault`

Reply:
At the moment, I don't see a way to create a default value based on the provided key. If there's no way to do so, it's not straightforward for me to understand how `defaultdict` can be helpful. (I'm specifically referring to the AuthorCompound class, which takes author as an argument in the constructor)